packages/database/src/core/WriteTree.ts

```diff
@@ -59,575 +59,745 @@ export interface WriteRecord {
+  completeServerChildren: ChildrenNode | null
+): ChildrenNode {
+  return writeTreeCalcCompleteEventChildren(
+    writeTreeRef.writeTree_,
+    writeTreeRef.treePath_,
+    completeServerChildren
+  ) as ChildrenNode;
+}
+
+/**
+ * Given that either the underlying server data has updated or the outstanding writes have updated, determine what,
+ * if anything, needs to be applied to the event cache.
+ *
+ * Possibilities:
+ *
+ * 1. No writes are shadowing. Events should be raised, the snap to be applied comes from the server data
+ *
+ * 2. Some write is completely shadowing. No events to be raised
+ *
+ * 3. Is partially shadowed. Events should be raised
+ *
+ * Either existingEventSnap or existingServerSnap must exist, this is validated via an assert
+ *
+ *
+ */
+export function writeTreeRefCalcEventCacheAfterServerOverwrite(
+  writeTreeRef: WriteTreeRef,
+  path: Path,
+  existingEventSnap: Node | null,
+  existingServerSnap: Node | null
+): Node | null {
+  return writeTreeCalcEventCacheAfterServerOverwrite(
+    writeTreeRef.writeTree_,
+    writeTreeRef.treePath_,
+    path,
+    existingEventSnap,
+    existingServerSnap
+  );
+}
+
+/**
+ * Returns a node if there is a complete overwrite for this path. More specifically, if there is a write at
+ * a higher path, this will return the child of that write relative to the write and this path.
+ * Returns null if there is no write at this path.
+ *
+ */
+export function writeTreeRefShadowingWrite(
+  writeTreeRef: WriteTreeRef,
+  path: Path
+): Node | null {
+  return writeTreeShadowingWrite(
+    writeTreeRef.writeTree_,
+    pathChild(writeTreeRef.treePath_, path)
+  );
+}
+
+/**
+ * This method is used when processing child remove events on a query. If we can, we pull in children that were outside
+ * the window, but may now be in the window
+ */
+export function writeTreeRefCalcIndexedSlice(
+  writeTreeRef: WriteTreeRef,
+  completeServerData: Node | null,
+  startPost: NamedNode,
+  count: number,
+  reverse: boolean,
+  index: Index
+): NamedNode[] {
+  return writeTreeCalcIndexedSlice(
+    writeTreeRef.writeTree_,
+    writeTreeRef.treePath_,
+    completeServerData,
+    startPost,
+    count,
+    reverse,
+    index
+  );
+}
+
+/**
+ * Returns a complete child for a given server snap after applying all user writes or null if there is no
+ * complete child for this ChildKey.
+ */
+export function writeTreeRefCalcCompleteChild(
+  writeTreeRef: WriteTreeRef,
+  childKey: string,
+  existingServerCache: CacheNode
+): Node | null {
+  return writeTreeCalcCompleteChild(
+    writeTreeRef.writeTree_,
+    writeTreeRef.treePath_,
+    childKey,
+    existingServerCache
+  );
+}
+
+/**
+ * Return a WriteTreeRef for a child.
+ */
+export function writeTreeRefChild(
+  writeTreeRef: WriteTreeRef,
+  childName: string
+): WriteTreeRef {
+  return new WriteTreeRef(
+    pathChild(writeTreeRef.treePath_, childName),
+    writeTreeRef.writeTree_
+  );
 }
 
 /**
```

Comment:
Even if you keep `WriteTree` as a class, `WriteTreeRef` should probably still be converted to an interface since it is only constructed in this module.

Reply:
Changed to an interface